balancing/balance_breaker.go

```diff
@@ -0,0 +1,615 @@
+	if state == closed {
+		if changed {
+			breaker.state = nil
+		}
+		return false
+	}
+	if state == halfopen {
+		if changed {
+			breaker.reset()
+		}
+		return false
+	}
+	return true
+}
+
+func (breaker *NodeBreaker) limitsExceeded() bool {
+	errorRate := breaker.errorRate()
+	if errorRate > breaker.rate {
+		breaker.openBreaker()
+		log.Debugf("Breaker: error rate exceeded %f", errorRate)
+		return true
+	}
+	percentile := breaker.timeData.Percentile(breaker.timeLimitPercentile)
+	if percentile > float64(breaker.callTimeLimit) {
+		breaker.openBreaker()
+		log.Debugf("Breaker: time percentile exceeded %f", percentile)
+		return true
+	}
+	return false
+}
+
+func (breaker *NodeBreaker) openBreaker() {
+	if breaker.state != nil {
+		return
+	}
+	breaker.state = newOpenStateTracker(
+		breaker.now(), breaker.closeDelay, breaker.maxDelay)
+}
+
+func (breaker *NodeBreaker) reset() {
+	breaker.timeData.Reset()
+	breaker.successData.Reset()
+}
+
+func (breaker *NodeBreaker) errorRate() float64 {
+	sum := breaker.successData.Sum()
+	count := float64(len(breaker.successData.values))
+	return sum / count
+}
+
+func newLenLimitCounter(retention int) *lengthDelimitedCounter {
+	return &lengthDelimitedCounter{
+		values: make([]float64, retention, retention),
+	}
+}
+
+type lengthDelimitedCounter struct {
+	values  []float64
+	nextIdx int
+	mx      sync.Mutex
+}
+
+// Add acumates new values
+func (counter *lengthDelimitedCounter) Add(value float64) {
+	counter.mx.Lock()
+	defer counter.mx.Unlock()
+	index := counter.nextIdx
+	counter.values[index] = value
+	counter.nextIdx = (counter.nextIdx + 1) % cap(counter.values)
+}
+
+// Sum returns sum of values
+func (counter *lengthDelimitedCounter) Sum() float64 {
+	sum := float64(0)
+	for _, v := range counter.values {
+		sum += v
+	}
+	return sum
+}
+
+// Percentile return value for given percentile
+func (counter *lengthDelimitedCounter) Percentile(percentile float64) float64 {
+	snapshot := make([]float64, len(counter.values))
+	copy(snapshot, counter.values)
+	sort.Float64s(snapshot)
+	pertcentileIndex := int(math.Floor(float64(len(snapshot)) * percentile))
+	return snapshot[pertcentileIndex]
+}
+
+func (counter *lengthDelimitedCounter) Reset() {
+	for idx := range counter.values {
+		counter.values[idx] = 0
+	}
+}
+
+type breakerState int
+
+const (
+	open     breakerState = 0
+	halfopen              = iota
+	closed                = iota
+)
+
+func newOpenStateTracker(start time.Time, changeDelay, maxDelay time.Duration) *openStateTracker {
+	return &openStateTracker{
+		lastChange:  start,
+		state:       open,
+		changeDelay: changeDelay,
+		maxDelay:    maxDelay,
+	}
+}
+
+type openStateTracker struct {
+	state          breakerState
+	lastChange     time.Time
+	changeDelay    time.Duration
+	maxDelay       time.Duration
+	closeIteration float64
+}
+
+func (tracker *openStateTracker) currentDelay() time.Duration {
+	multiplier := int(math.Pow(2, tracker.closeIteration))
```

Comment:
`closeIteration` is never decremented/reset, even if a backend is healthy for a long time. 
An interesting case popped up in my head.
> initially the backend is not doing so well nad its `closeIteration` gets incremented until it reaches some high value
> the backend becomes healthy and stays healthy for a long time, but it's `closeIteration` is already big
> the backend misbehaves for a brief second and gets taken out of the pool of healthy nodes for a long time, due to it's `closeIteration`
> we wait a long time until we plug the backend back in, even though it' already healthy and might provide us with the best timings